src/v/storage/disk_log_impl.cc

```diff
@@ -587,7 +587,20 @@ ss::future<bool> disk_log_impl::sliding_window_compact(
             cfg.asrc->check();
         }
         if (seg->offsets().base_offset > map.max_offset()) {
-            break;
+            // The map was built from newest to oldest segments within this
+            // sliding range. If we see a new segment whose offsets are all
+            // higher than those indexed, it may be because the segment is
```

Comment:
> [0, 10)[10, 20)[20, 25)
> The compacted indices only contains data keys, so the key offset map
will only reflect data up to offset 20.

Here in the code iteration is from lowest offset to highest offset. I'm trying to figure out where the iteration would stop. Presumably the largest off in the index could be 19, so iteration would stop on the [20,25) segment?

Does this same basic principle apply to other situations that don't involve segments with all non-data batches? For example if the index fills up, could we have the same effect but even if the segment contained data batches?

Reply:
>Presumably the largest off in the index could be 19, so iteration would stop on the [20,25) segment?

Iteration would continue to [20, 25) and mark it compacted, since it none of the batches in the segment would be affected by deduplication with the offset map.

>For example if the index fills up, could we have the same effect but even if the segment contained data batches?

Because the key offset map is built with a constant range of segments, and because we build our index from newest to oldest and only index data batches, if we ever see a segment that gets into this if statement, either that segment and all newer segments did not have any keys, or the segment isn't in our constant range of segments (contradiction!)